### PR TITLE
[WIP] Goal progress tracking

### PR DESCRIPTION
**Pull request by andycmaj**
SPEC REVIEW

### CODE REVIEW

File: specs/52-goal_progress.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# [Goal Progress Tracking](https://github.com/asynchronous-dev/botany/issues/52)
```
Comment (andycmaj) [inline specs/52-goal_progress.md]: @davemeyer am I over complicating things? I feel like metrics tracking and triggering is inherently a hard problem. 

does the whole `Goals can be _RELEVANT TO_ either **Actor-specific metrics** or **Entity-specific metrics**. ` thing make sense to you? and do i seem to be modeling this correctly with `userGoalMetrics`?

File: specs/37-metrics.md [hunk 2]
```diff
@@ -1,35 +1,42 @@
...
-  - look up metrics when consuming `botany_activity`
-    - `activity → actor → userGoals → userGoalMetrics → metrics`
+- `MetricProcessor` js modules (`CodeReviewHotness`)
+  - `Metric` calculation and id generation/lookup
+  - Gets relevant `UserGoals` with `UserGoalMetrics` to track threshold/progress
+- relation table linking `UserGoal` with `Metric`
+  - create `Metric`s lazily when processing `botany_activity`
+    - handle activity
+    - Create or Update `Metric` related to this activity/`MetricProcessor` pair.
+    - look up `userGoals` for `actor`
```
Comment (davemeyer) [inline specs/37-metrics.md]: so walk me through an example for `pr_comments_added`. that would likely result in metrics being updated for the person who left the comments. maybe one for the person who wrote the pr. maybe something for the whole team? and then something for the pr `Entity`. is the flow: `raw activity -> lookup all goal types associated with this event -> lookup all users with this goal (where does the entity come in?) -> fire events for each `actor` -> each event is handled by the metrics processor`

File: specs/52-goal_progress.md (hunk 1 above)
Comment (davemeyer) [reply specs/52-goal_progress.md]: yeah breaking things into actor-specific and entity-specific makes sense. it is going to result in "duplicate-ish" data. but it seems like a rational way to model it. see my above question though so i can make sure i'm fully on the same page with the steps in the process

File: specs/52-goal_progress.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+The `subjectId` for these metrics would be the `Actor` id.
+
+### Entity-specific Metrics
+
+Metrics that are not tied to a specific Botany `Actor`. They could, for example track values related to a Github Code Review or a JIRA Epic.
+
+## Goals
+
+Goals can be _RELEVANT TO_ either **Actor-specific metrics** or **Entity-specific metrics**. For example, `goal:code_review_participation` has a **MetricThreshold-triggered nudge** for the `metric:code_review_hotness`. This metric is **Entity-specific** (the metric subject is the code review itself). But `goal:code_review_participation` also tracks progress based on an **Actor-specific** metric, `metric:weekly_code_review_participation`, with `subjectId=someActor`.
```
Comment (davemeyer) [inline specs/52-goal_progress.md]: how do you plan to identify the `Entities` for entity-specific metrics?

File: specs/37-metrics.md (hunk 2 above)
Comment (andycmaj) [reply specs/37-metrics.md]: not exactly. this is where the actor/goal-specific vs entity-specific metrics come in.
the entity-specific metric example is the one i've built in my other PR, `code_review_hotness`. 

https://github.com/asynchronous-dev/botany/blob/8d29eb7886fe2fa481b22e390169d7010d162342/consumers/src/handlers/processMetrics/CodeReviewHotness.ts#L164

the flow here is:

* handle `botany_activity`: `code_review_discussed` in this case.
* call all relevant metrics handlers (eg. `CodeReviewHotness.ts`) mapped to the `botany_activity` `type`. (so `code_review_discussed` -> `[CodeReviewHotness]`)
  * hotness is an entity-specific metric, so it's tracked as a single metric with `subjectId=<code_review_uri>`.
  * if none exists yet create the metric now. and set the value to `commentCount:1`.
  * then find all relevant `userGoals` so we can nudge (nudging only for this metric. there aren't any actor-specific metrics here yet, but entity-specific metrics can trigger nudges for relevant goals.
  * for each `userGoal` that has a `goal` type that cares about metrics of type `code_review_hotness`  and with subject `<some code review by anyone on same team as user with the userGoal>`, look at the metric threshold for `code_review_hotness` and nudge each `user` with a `userGoalMetric` with a crossed threshold.

so TL;DR; how i THINK it's gonna work is 
* entity-specific metrics will first update/create the metric, then find users with relevant `userGoal`s to nudge.
* actor-specific metrics will first find all relevant `userGoalMetrics`, update the metric, update `userGoal` progress, and nudge if necessary

will be clearer once i've created a `userGoal` that's linked to actor-specific metrics that have progress that gets updated... there's not really a relevant "goal progress" to be updated based on `code_review_hotness`.

Comment (andycmaj) [reply specs/37-metrics.md]: > walk me through an example for pr_comments_added. that would likely result in metrics being updated for the person who left the comments

and yeah... that was gonna be the first actor-specific metric i was going to add after this spec gets discussed. 

i'll add a new metric for something like `weekly_code_review_participation` and `userGoalMetrics` for `goal:code_review_participation` -> `userGoal:andy-code_review_participation` -> `userGoalMetric:andy-code_review_participation-weekly_code_review_participation`.

and flow will be something like...

* handle botany_activity `code_review_discussed` with actor `andy`.
* map `code_review_discussed` -> `WeeklyCodeReviewParticipation`
* find `userGoals` for actor, update `weekly_code_review_participation` metric for relevaant goal
* send positive feedback at some point

File: specs/52-goal_progress.md (hunk 3 above)
Comment (andycmaj) [reply specs/52-goal_progress.md]: that's the `subjectId`. 

something like 
```
 const metrics = await client.query.metric({
    where: {
      name: { _eq: 'code_review_hotness' },
      subjectId: { _eq: activity.subject },
    },
  });
```

in the case of a code review, it's the github pull request uri. actually shoudl always be a URI if possible

if not `http://github.com/...`, then `botany://actor/1234` for internal entities. but not sure if we'll need that

Comment (andycmaj) [reply specs/52-goal_progress.md]: for example, `github/pullRequest` enricher sets this here: https://github.com/asynchronous-dev/botany/blob/8d29eb7886fe2fa481b22e390169d7010d162342/consumers/src/handlers/enrichRawActivity/github/pullRequest.ts#L35

**Comment by andycmaj** [comment]
here's an example pattern https://github.com/Urigo/WhatsApp-Clone-Client-React/blob/8d53497dda75bd48300258c30693487d526a8a7b/src/services/auth.service.tsx#L14
